### PR TITLE
Add MultistreamProtocol with ability to inject Stream/Connection handlers in two points 

### PR DESCRIPTION
**Pull request by Nashatyrev**
Based on top of PR #170, but can be viewed as a single change since removes some updates made in #170

A bit more refactoring to make Libp2p more open and customizable: 

- Add `MultistreamProtocol` aka factory for `Multistream` instances 
- Add `StreamMuxerProtocol` aka factory for `StreamMuxer` instances 
- Add `MultistreamProtocolDebug` interface with ability to add `Stream/ConnectionVisitor` before and after multistream negotiation 
- With `MultistreamProtocolDebug` replace `StreamVisitor`s and `ConnectionUpgrader` debug handlers
- Add `Host.DebugBuilder` `streamPreHandler` and `streamHandler`: the former may handle a `Stream` right after creation (with all preliminary multistream negotiations), the latter handles a Stream right after the protocol negotiated (this significantly simplifies specific protocol listeners/interceptors)

### CODE REVIEW

**Review by mbaxter — APPROVED** [review]

File: src/main/kotlin/io/libp2p/core/dsl/Builders.kt [hunk 1]
```diff
@@ -58,6 +63,10 @@ open class Builder {
     protected open val connectionHandlers = ConnectionHandlerBuilder()
     protected open val network = NetworkConfigBuilder()
     protected open val debug = DebugBuilder()
+    var multistreamProtocol: MultistreamProtocol = MultistreamProtocol_v_1_0_0
```
[the remark this answers is not in the document]
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/core/dsl/Builders.kt:66]: `var` is intentional. One may want to change the `multistream` protocol in `Builder`

File: src/main/kotlin/io/libp2p/core/multistream/MultistreamProtocol.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    private val preHandler: P2PChannelHandler<*>? = null,
+    private val postHandler: P2PChannelHandler<*>? = null
+) : MultistreamProtocolDebug {
+
+    override val version = "1.0.0"
+
+    override fun <TController> create(bindings: List<ProtocolBinding<TController>>) =
+        MultistreamImpl(bindings, preHandler, postHandler)
+
+    override fun copyWithHandlers(
```
[the remark this answers is not in the document]
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/core/multistream/MultistreamProtocol.kt:36]: In this implementation there is actually nothing to copy

[the remark this answers is not in the document]
Comment (mbaxter) [reply src/main/kotlin/io/libp2p/core/multistream/MultistreamProtocol.kt:36]: The constructor for `MultistreamProtocolDebug_v_1_0_0` is public though - so couldn't we be copying from an instance that already has pre- and post- handlers?

[the remark this answers is not in the document]
Comment (Nashatyrev) [reply src/main/kotlin/io/libp2p/core/multistream/MultistreamProtocol.kt]: Oh, I think I got what you mean. Please check this commit: 976e896

[the remark this answers is not in the document]
Comment (mbaxter) [reply src/main/kotlin/io/libp2p/core/multistream/MultistreamProtocol.kt]: Looks good 👍

**Comment by Nashatyrev** [comment]
Here is the Teku adoption changes: https://github.com/Nashatyrev/artemis/commit/609a8a181308d6a2d0503b767e212cdd22990e0c

**Pull request merged**